crates/router/src/connector/globalpay/transformers.rs

```diff
@@ -72,21 +79,70 @@ impl TryFrom<&types::PaymentsCancelRouterData> for GlobalpayPaymentsRequest {
 }
 
 pub struct GlobalpayAuthType {
-    pub api_key: String,
+    pub app_id: String,
+    pub key: String,
 }
 
 impl TryFrom<&types::ConnectorAuthType> for GlobalpayAuthType {
     type Error = error_stack::Report<errors::ConnectorError>;
     fn try_from(auth_type: &types::ConnectorAuthType) -> Result<Self, Self::Error> {
         match auth_type {
-            types::ConnectorAuthType::HeaderKey { api_key } => Ok(Self {
-                api_key: api_key.to_string(),
+            types::ConnectorAuthType::BodyKey { api_key, key1 } => Ok(Self {
+                app_id: api_key.to_string(),
+                key: key1.to_string(),
             }),
             _ => Err(errors::ConnectorError::FailedToObtainAuthType.into()),
         }
     }
 }
 
+impl From<GlobalpayRefreshTokenErrorResponse> for types::ErrorResponse {
+    fn from(item: GlobalpayRefreshTokenErrorResponse) -> Self {
+        Self {
+            code: item.error_code,
+            message: item.detailed_error_description,
+            reason: None,
+        }
+    }
+}
+
+impl TryFrom<GlobalpayRefreshTokenResponse> for types::AccessToken {
+    type Error = error_stack::Report<errors::ParsingError>;
+
+    fn try_from(item: GlobalpayRefreshTokenResponse) -> Result<Self, Self::Error> {
+        Ok(Self {
+            token: item.token,
+            expires: item.seconds_to_expire,
+        })
+    }
+}
+
+impl TryFrom<&types::RefreshTokenRouterData> for GlobalpayRefreshTokenRequest {
+    type Error = error_stack::Report<errors::ConnectorError>;
+
+    fn try_from(item: &types::RefreshTokenRouterData) -> Result<Self, Self::Error> {
+        let globalpay_auth = GlobalpayAuthType::try_from(&item.connector_auth_type)
+            .change_context(errors::ConnectorError::FailedToObtainAuthType)
+            .attach_printable("Could not convert connector_auth to globalpay_auth")?;
+
+        let current_time = date_time::now().to_string();
+        let nonce_with_api_key = format!("{}{}", current_time, globalpay_auth.key);
+        let secret_vec = crypto::Sha512
+            .generate_digest(nonce_with_api_key.as_bytes())
+            .change_context(errors::ConnectorError::RequestEncodingFailed)
+            .attach_printable("error creating request nonce")?;
+
+        let secret = hex::encode(secret_vec);
```

Comment:
Have they specified that the nonce must be the current time, and that the format/contents of the secret doesn't matter?

Preferably, if you need random data, you should be using one of the RNGs from the `rand` crate.

Reply:
No they haven't specified that, but isn't using current time faster than creating random numbers?

Reply:
If I remember correctly, there was a discussion about this involving @vspecky. If you need random numbers, then use random numbers. Don't substitute them with timestamp.

Reply:
Yes, using system time can have two disadvantages :-
- It can be slower than your RNG since it needs to make a syscall (usually insignificant)
- The system time on the server does not move forward at a constant speed. When the server is synchronizing time with a reference server using NTP, the system time may speed up, slow down or even jump around, so it's not guaranteed to be completely random. I don't know if in this case it could be a problem but generally with cryptography, it's better if we go with an RNG since it will generally be "more random" than system time.